stdlib/experimental/dynamic/dynamic.go

```diff
@@ -21,10 +27,94 @@ var dynamicConv = values.NewFunction(
 			return nil, err
 		}
 
+		// Nothing to do if the incoming value is already a dynamic
+		if v.Type().Nature() == semantic.Dynamic {
+			v.Retain()
+			return v, nil
+		}
+
 		// FIXME(onelson): do we need to special case nulls here?
 		//   Ex: if the input is null, should the output be null or a
 		//   Dynamic with a null in it? My sense is the latter.
 		return values.NewDynamic(v), nil
 	},
 	false,
 )
+
+var asArray = values.NewFunction(
+	"asArray",
+	runtime.MustLookupBuiltinType("experimental/dynamic", "asArray"),
+	func(ctx context.Context, args values.Object) (values.Value, error) {
+		arguments := interpreter.NewArguments(args)
+		v, err := arguments.GetRequired("v")
+		if err != nil {
+			return nil, err
+		}
+
+		if v.IsNull() {
+			panic("TODO")
+		}
+
+		d := v.Dynamic()
+		inner := d.Inner()
+		if inner.Type().Nature() != semantic.Array {
+			return nil, errors.Newf(codes.Invalid, "unable to convert %s to array", inner.Type())
+		}
+		arr := inner.Array()
+		elmType, err := arr.Type().ElemType()
+		if err != nil {
+			return nil, err
+		}
+
+		// The contract for this function says it accepts a dynamic and produces
+		// an array of dynamic.
+		// Note that just because we've verified the argument `v` was a dynamic
+		// value holding an array, it doesn't mean the elements inside that array
+		// are wrapped in dynamic.
+		// Therefore, check to see if the elements are dynamic and wrap them if they aren't.
+		if elmType.Nature() == semantic.Dynamic {
+			arr.Retain()
+			return arr, nil
+		} else {
+			elems := make([]values.Value, arr.Len())
+			for i := 0; i < arr.Len(); i++ {
+				v := arr.Get(i)
+				v.Retain()
+				elems[i] = values.NewDynamic(v)
+			}
+			dynArr := values.NewArrayWithBacking(semantic.NewArrayType(semantic.NewDynamicType()), elems)
+			return dynArr, nil
+		}
+	},
+	false,
+)
```

Comment:
Seeing calls to `Retain` here makes me wonder if we would catch bugs related to reference counts in our acceptance tests? I know that we can catch them with Go unit tests by using a special allocator and assert there is no more allocated bytes at the end of the test. We should do something similar when we run our acceptance tests. I don't know if we do that or not.

Reply:
This is a good idea. Right now a lot of the vectorization tests are partially duplicated in Go just to verify the memory accounting. We currently don't.

I don't know that it'll be a big effort, but I'd like to file a separate issue to set that up. I suspect the smaller lift is just to add a set of Go tests, but it'd be great to not have to do that.

Reply:
I filed this as https://github.com/influxdata/flux/issues/5251